reference/test/matrix/csr_kernels.cpp

```diff
@@ -1760,4 +1764,85 @@ TYPED_TEST(Csr, CanGetSubmatrix2)
+{
+    using Vec = typename TestFixture::Vec;
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    auto mat = gko::initialize<Mtx>(
+        {
+            I<T>{1.0, 3.0, 4.5, 0.0, 2.0},   // 0
+            I<T>{1.0, 0.0, 4.5, 7.5, 3.0},   // 1
+            I<T>{0.0, 3.0, 4.5, 0.0, 2.0},   // 2
+            I<T>{0.0, -1.0, 2.5, 0.0, 2.0},  // 3
+            I<T>{1.0, 0.0, -1.0, 3.5, 1.0},  // 4
+            I<T>{0.0, 1.0, 0.0, 0.0, 2.0},   // 5
+            I<T>{0.0, 3.0, 0.0, 7.5, 1.0}    // 6
+        },
+        this->exec);
+    ASSERT_EQ(mat->get_num_stored_elements(), 23);
+    {
+        SCOPED_TRACE("Small square 2x2");
+        auto row_set = gko::IndexSet<index_type>(this->exec, {0, 1});
+        auto col_set = gko::IndexSet<index_type>(this->exec, {0, 1});
+        auto sub_mat1 = mat->create_submatrix(row_set, col_set);
+        auto ref1 =
+            gko::initialize<Mtx>({I<T>{1.0, 3.0}, I<T>{1.0, 0.0}}, this->exec);
+
+        GKO_EXPECT_MTX_NEAR(sub_mat1.get(), ref1.get(), 0.0);
+    }
+
+    {
+        SCOPED_TRACE("Non-square 4x2");
+        auto row_set = gko::IndexSet<index_type>(this->exec, {1, 2, 3, 4});
+        auto col_set = gko::IndexSet<index_type>(this->exec, {1, 3});
+        auto sub_mat1 = mat->create_submatrix(row_set, col_set);
+        auto ref1 = gko::initialize<Mtx>(
+            {I<T>{0.0, 7.5}, I<T>{3.0, 0.0}, I<T>{-1.0, 0.0}, I<T>{0.0, 3.5}},
+            this->exec);
+
+        GKO_EXPECT_MTX_NEAR(sub_mat1.get(), ref1.get(), 0.0);
+    }
+
+    {
+        SCOPED_TRACE("Square 3x3");
+        auto row_set = gko::IndexSet<index_type>(this->exec, {1, 3, 4});
+        auto col_set = gko::IndexSet<index_type>(this->exec, {1, 3, 0});
+        auto sub_mat1 = mat->create_submatrix(row_set, col_set);
+        auto ref1 = gko::initialize<Mtx>(
+            {I<T>{1.0, 0.0, 7.5}, I<T>{0.0, -1.0, 0.0}, I<T>{1.0, 0.0, 3.5}},
+            this->exec);
+
+        GKO_EXPECT_MTX_NEAR(sub_mat1.get(), ref1.get(), 0.0);
+    }
+
+    {
+        SCOPED_TRACE("Square 4x4");
+        auto row_set = gko::IndexSet<index_type>(this->exec, {1, 4, 5, 6});
+        auto col_set = gko::IndexSet<index_type>(this->exec, {4, 3, 0, 1});
+        auto sub_mat1 = mat->create_submatrix(row_set, col_set);
+        auto ref1 = gko::initialize<Mtx>({I<T>{1.0, 0.0, 7.5, 3.0},   // 1
+                                          I<T>{1.0, 0.0, 3.5, 1.0},   // 4
+                                          I<T>{0.0, 1.0, 0.0, 2.0},   // 5
+                                          I<T>{0.0, 3.0, 7.5, 1.0}},  // 6
+                                         this->exec);
+
+        GKO_EXPECT_MTX_NEAR(sub_mat1.get(), ref1.get(), 0.0);
+    }
+
+    {
+        SCOPED_TRACE("Non Square 2x4");
+        auto row_set = gko::IndexSet<index_type>(this->exec, {5, 6});
+        auto col_set = gko::IndexSet<index_type>(this->exec, {4, 3, 0, 1});
```

Comment:
if we set it is sorted, we will get the permutated one?
sub_col 0 <-> src_col 4
sub_col 1 <-> src_col 3
sub_col 2 <-> src_col 0
sub_col 3 <-> src_col 1

Reply:
No, the index_set always sorts its input and then stores them. I wanted to make sure that with unsorted inputs to index_set, this algorithm still works

Reply:
maybe add a comment here explaining that detail?